nano/lib/timer.hpp

```diff
@@ -0,0 +1,250 @@
+	 * multiple times (e.g. in a loop).
+	 * @return duration
+	 */
+	UNIT pause ()
+	{
+		++measurements;
+		return stop ();
+	}
+
+	/**
+	 * Stop timer
+	 * @return duration
+	 */
+	UNIT stop ()
+	{
+		assert (state == nano::timer_state::started);
+		state = nano::timer_state::stopped;
+
+		auto end = std::chrono::high_resolution_clock::now ();
+		ticks += std::chrono::duration_cast<UNIT> (end - begin);
+		return ticks;
+	}
+
+	/** Returns the duration in UNIT since the timer was last started. */
+	UNIT since_start ()
+	{
+		auto end = std::chrono::high_resolution_clock::now ();
+		return std::chrono::duration_cast<UNIT> (end - begin);
+	}
+
+	/** Returns true if the timer was last started longer than \p duration_a units ago*/
+	bool after_deadline (UNIT duration_a)
+	{
+		auto end = std::chrono::high_resolution_clock::now ();
+		return std::chrono::duration_cast<UNIT> (end - begin) > duration_a;
+	}
+
+	/** Returns true if the timer was last started shorter than \p duration_a units ago*/
+	bool before_deadline (UNIT duration_a)
+	{
+		auto end = std::chrono::high_resolution_clock::now ();
+		return std::chrono::duration_cast<UNIT> (end - begin) < duration_a;
+	}
+
+	/** Stop timer and write measurements to \p output_a */
+	void stop (std::string & output_a)
+	{
+		std::ostringstream stream;
+		stop (stream);
+		output_a = stream.str ();
+	}
+
+	/** Stop timer and write measurements to \p stream_a */
+	void stop (std::ostream & stream_a)
+	{
+		stop ();
+		print (stream_a);
+	}
+
+	/** Print measurements to the \p stream_a */
+	void print (std::ostream & stream_a)
+	{
+		if (ticks >= minimum)
+		{
+			// Print cumulative children first. Non-cumulative children prints directly.
+			for (auto & child : children)
+			{
+				if (child.measurements > 0)
+				{
+					child.print (stream_a);
+				}
+			}
+
+			auto current_parent = parent;
+			while (current_parent)
+			{
+				stream_a << parent->desc << ".";
+				current_parent = current_parent->parent;
+			}
+
+			stream_a << desc << ": " << ticks.count () << ' ' << unit ();
+			if (measurements > 0)
+			{
+				stream_a << " (" << measurements << " measurements, " << std::setprecision (2) << std::fixed << static_cast<double> (ticks.count ()) / static_cast<double> (measurements) << ' ' << unit () << " avg)";
+			}
+			stream_a << std::endl;
+		}
+	}
+
+	/** Returns the SI unit string */
+	std::string unit () const
+	{
+		UNIT u;
+		return typed_unit (u);
+	}
+
+	nano::timer_state current_state () const
+	{
+		return state;
+	}
+
+private:
+	timer * parent{ nullptr };
+	std::vector<timer> children;
+	nano::timer_state state{ nano::timer_state::stopped };
+	std::string desc;
+	std::chrono::time_point<std::chrono::high_resolution_clock> begin;
+	UNIT ticks{ 0 };
+	UNIT minimum{ UNIT::zero () };
+	unsigned measurements{ 0 };
+
+	template <typename U = UNIT>
+	std::string typed_unit (typename std::enable_if<std::is_same<U, std::chrono::nanoseconds>::value, std::chrono::nanoseconds>::type dummy) const
```

Comment:
I quite like this approach. There is a helper alias [std::enable_if_t](https://en.cppreference.com/w/cpp/types/enable_if) in C++14 to shorten this syntax. You can also avoid the "dummy" variable by using it in the return type, i.e:
`std::enable_if_t<std::is_same<U, std::chrono::microseconds>::value, std::string> typed_unit () { return "microseconds";}`

Or even better inside the template argument to clean up the function interface:
```
template <typename U,
          std::enable_if_t<std::is_same<U, std::chrono::nanoseconds>::value>* = nullptr >
std::string typed_unit ()
{
	return "nanoseconds";
}

```

You could also create your own template alias to reduce some of the duplication and make it a bit tidier, something like this:

```
template<typename U, typename V>
using MatchesType = std::enable_if_t<std::is_same<U, V>::value>;

template <typename U, MatchesType<U, std::chrono::nanoseconds>* = 0>
std::string typed_unit ()
{
	return "nanoseconds";
}

template <typename U, MatchesType<U, std::chrono::microseconds>* = 0>
std::string typed_unit ()
{
	return "microseconds";
}

...
```

Then your unit () function would just contain:
 `return typed_unit<UNIT>();`

Reply:
Yeah, the C++14 ...if_t is nicer - fixed.

Reply:
Looks good! In C++17 there is also a helper variable template **std::is_same_v<U, V**>as well